### PR TITLE
backport: bitcoin#18531, #20012, #21035, #21572, #21574 (rpc command) and related fixes

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
Backports from bitcoin v22 rpc command related

## What was done?
See commits for backports.
Also:
 - refactored and significantly simplified implementation of composite commands
 - added missing changes from bitcoin#19250
 - fix  help message for rpc `help` - all subcommands support it now
 - add missing client's argument parsing for RPC commands
 - CoreContext uses std::monostate instead nullopt which is best-practice for std::variant

## How Has This Been Tested?
Run unit/functional tests.
Checked autocomplete for various commands
Checked help for various commands

## Breaking Changes
n/a

## Checklist:
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by knst** [comment]
In theory, this PR can be split to 3...7 PRs, but they can not be reviewed and merged in parallel, they should be done in the specific order one-by-one. So, that's an one PR. 

For example, it can be split like that:

1st round: Prior fixes  (1PR):
 - refactor: use monostate instead std::optional in CoreContext
 - fix: missing changes from bitcoin#19250
 - feat: fix help message - all subcommands support it now!
 - fix: add missing client's argument parsing for RPC commands

2nd round: 3 Backports (1 PR):
 - Merge #18531: rpc: remove deprecated CRPCCommand constructor
 - Merge #20012: rpc: Remove duplicate name and argNames from CRPCCommand
 - chore: add TODO to make client parsing for composite commands
 
3rd round: Follow-up refactoring (1 PR):
 - refactor: simplify implementation of RPC composite commands

4th round: And finally 3 remaining backports (1 PR):
 - Merge #21035: Remove pointer cast in CRPCTable::dumpArgMap
 - Merge #21572: Fix wrong wallet RPC context set after #21366
 - Merge #21574: Drop JSONRPCRequest constructors after #21366

File: src/node/interfaces.cpp [hunk 1]
```diff
@@ -584,11 +585,11 @@ class NodeImpl : public Node
         if (context) {
             m_context_ref = *context;
         } else {
-            m_context_ref = std::nullopt;
+            m_context_ref = std::monostate{};
         }
     }
     NodeContext* m_context{nullptr};
-    CoreContext m_context_ref{std::nullopt};
+    CoreContext m_context_ref{std::monostate{}};
```
Comment (UdjinM6) [inline src/node/interfaces.cpp]: d34ab741269e0042584584d022c4dce7a7f7a596: monostate makes it a default-constructed variant which holds a value of its first alternative, no need to initialize it explicitly anymore.
```suggestion
    CoreContext m_context_ref;
```

File: src/rpc/request.h [hunk 2]
```diff
@@ -36,18 +36,7 @@ class JSONRPCRequest
...
-
-    //! Initializes request information from another request object and the
-    //! given context. The implementation should be updated if any members are
-    //! added or removed above.
-    JSONRPCRequest(const JSONRPCRequest& other, const CoreContext& context)
-        : id(other.id), strMethod(other.strMethod), params(other.params), mode(other.mode), URI(other.URI),
-          authUser(other.authUser), peerAddr(other.peerAddr), context(context)
-    {
-    }
+    CoreContext context{std::monostate()};
```
Comment (UdjinM6) [inline src/rpc/request.h:39]: 21574: same

File: src/context.h [hunk 3]
```diff
@@ -6,7 +6,6 @@
 #define BITCOIN_CONTEXT_H
 
 #include <variant>
```
Comment (UdjinM6) [inline src/context.h:8]: d34ab741269e0042584584d022c4dce7a7f7a596: you need `<functional>` for `reference_wrapper`
```suggestion
#include <variant>
#include <functional>
```

**Review by UdjinM6 — COMMENTED** [review]
few notes

**Review by UdjinM6 — APPROVED** [review]
utACK bfc083e9b792d9c342b336a121de046f668f9799

**Pull request merged**